### PR TITLE
HeadTracker sampled heads

### PR DESCRIPTION
**Pull request by PiotrTrzpil**
In case of fast chains, doing work on every new head is a performance bottleneck and is not really needed. Services that would benefit:
EthConfirmer
GasUpdater
PromReporter
BalanceChecker

Interval could be configurable by HeadTrackerSamplingInterval

After going through the usages of latest head, it should not cause any unintended consequences, that some heads will be skipped:

EthConfirmer 
 - used for updating broadcast_before_block_num in eth_tx_attempts table (SetBroadcastBeforeBlockNum)
 - used as cutoff for marking txs as errored (current block num - finality depth) (markOldTxesMissingReceiptAsErrored)
 - used for finding a range of txs requiring gas bump (FindEthTxsRequiringGasBump)
 - chain is used to check if given receipt exist in it, and is rebroadcasted otherwise (EnsureConfirmedTransactionsInLongestChain)
 - reaper: used as part of a lower bound of transactions to keep (current block num - finality depth) (ReapEthTxes)

PromReporter
 - used to calculate num of blocks since the earliest one with an unconfirmed tx (reportMaxUnconfirmedBlocks)

GasUpdater 
 - used to calculate a range of blocks to fetch, trailing N behind the latest block num (FetchBlocks)
 - chain is used to check which blocks we already have in memory (FetchBlocks)

Keeper - Upkeep Executer
 - used to select a range of upkeeps below a certain head count before latest (EligibleUpkeeps)
 - used to mark upkeep registrations with last run block height (SetLastRunHeightForUpkeepOnJob)

LogBroadcaster
 - used to delete older logs that are below the kept range - more than N blocks below latest (getLogsToSend)
 - used for checking if given log is confirmed enough to be sent (sendLog)
 - chain is stored for each listener to check for reorgs (in case of a reorg, the lowestAllowedBlockNumber to send, is reset) (applyListenerInfoUpdates)

RunManager 
 - used for setting task runs confirmations (ResumeAllPendingNextBlock)

### CODE REVIEW

File: core/services/head_tracker.go [hunk 1]
```diff
@@ -289,8 +292,37 @@ func (ht *HeadTracker) disconnect() {
...
+		case <-ht.done:
+			return
+		case <-debounceHead.C:
+			item, exists := ht.samplingMB.Retrieve()
+			if !exists {
+				continue
+			}
+			head, ok := item.(models.Head)
+			if !ok {
+				logger.Errorf("expected `models.Head`, got %T", item)
```
Comment (samsondav) [inline core/services/head_tracker.go:315]: I think it's reasonable to panic here

File: core/services/head_tracker_test.go [hunk 2]
```diff
@@ -754,7 +754,6 @@ func (c *blockingCallback) Connect(bn *models.Head) error {
 func (c *blockingCallback) Disconnect() {
 }
 
-// OnNewLongestChain increases the OnNewLongestChainCount count by one
```
Comment (samsondav) [inline core/services/head_tracker_test.go:757]: Lol this comment was total garbage

File: core/services/prom_reporter.go [hunk 3]
```diff
@@ -97,6 +125,36 @@ func (pr *promReporter) Disconnect() {
...
+
+		case <-pr.chStop:
+			return
+		}
+	}
+}
+
+func (pr *promReporter) reportMetrics(head models.Head) {
+
+	ctx := context.Background()
```
Comment (samsondav) [inline core/services/prom_reporter.go:157]: Perhaps we could use the passed in context?

**Review by samsondav — APPROVED** [review]
Really nice

File: core/services/prom_reporter.go (hunk 3 above)
Comment (PiotrTrzpil) [reply core/services/prom_reporter.go:157]: yup, passed a context from chStop now

File: core/services/head_tracker.go (hunk 1 above)
Comment (PiotrTrzpil) [reply core/services/head_tracker.go:315]: cool, changed also 2 other places

**Review by samsondav — APPROVED** [review]

**Pull request merged**